### PR TITLE
[jit] Return the compiled method and an unbox trampoline from ldvirtftn when calling ldvirtftn on a valuetype method.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/dotnet/runtime/issues/34379.

### CODE REVIEW

**Comment by fanyang-mono** [comment]
Two coreclr tests related to this issue need to be enabled as well.
Please refer to my change to src/coreclr/tests/issues.targets here
https://github.com/dotnet/runtime/pull/35766/files

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
This seems to work now.

**Comment by vargaz** [comment]
To recap, the problem with the previous patch was that mono_delegate_ctor () would need to do an address->method reverse lookup, and it couldn't do it because address was pointing to an unbox trampoline, which was not registered in the JIT info tables.

File: mono/mini/jit-icalls.c [hunk 1]
```diff
@@ -93,8 +94,28 @@ ldvirtfn_internal (MonoObject *obj, MonoMethod *method, gboolean gshared)
 	}
 
 	/* An rgctx wrapper is added by the trampolines no need to do it here */
+	gboolean need_unbox = m_class_is_valuetype (res->klass) && !m_class_is_valuetype (method->klass);
+	if (need_unbox && !mono_use_interpreter) {
```
Comment (BrzVlad) [inline mono/mini/jit-icalls.c:98]: Is `!mono_use_interpreter` meant to fix something or is it just a left over from the initial PR ?

Comment (vargaz) [reply mono/mini/jit-icalls.c:98]: Will investigate.

Comment (fanyang-mono) [reply mono/mini/jit-icalls.c:98]: I did some test with or without this interpreter check and added my comment to the initial PR. So this check probably needs to be removed and something needs to be fixed for failing test. I am not sure if we should have a follow-up PR for that issue or have it fixed in this PR as well. That problem is related to fullAOT+interp.

**Comment by SamMonoRT** [comment]
You need to somehow port this to dotnet\runtime and update the issues.target file to run these changes and tests enabled and make sure those pass.

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/mini/jit-icalls.c (hunk 1 above)
Comment (vargaz) [reply mono/mini/jit-icalls.c]: The problem doesn't seem to be present any more.

File: mono/mini/aot-runtime.c [hunk 2]
```diff
@@ -6050,6 +6052,14 @@ mono_aot_get_unbox_trampoline (MonoMethod *method, gpointer addr)
 		return mono_aot_get_unbox_arbitrary_trampoline (addr);
 	}
 
+	if (!amodule->unbox_tramp_per_method) {
+		gpointer arr = g_new0 (gpointer, amodule->info.nmethods);
+		mono_memory_barrier ();
+		amodule->unbox_tramp_per_method = (gpointer*)arr;
```
Comment (BrzVlad) [inline mono/mini/aot-runtime.c:6058]: We should CAS here and free the old array if somebody else beat us to publishing the cache array. I'm not quite sure whether it is a problem if we register twice a tramp info (and twice a jit info) for the same code. We might have to lock and check if code was already published, and only register the tramp info once. I guess we would hav had this problem already for other trampolines if that's the case. Can't seem to find right now where we are handling concurrent compilations of the same methods with jit, for reference.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**